### PR TITLE
Upgrade active_record to 6

### PR DESCRIPTION
**Pull request by khaleksa**
## Description
Add support for Rails 6.

## Testing
The arel_toolkit is tested against Rails 6 with the help of `appraisals` gem with configuration:
```
appraise 'rails-6.0' do
  gem 'rails', '~> 6.0.0'
end
```

Preparation before testing:
1 - install ruby
2 - bin/setup
4 - bundle exec rake clean
5 - bundle exec rake compile

Install the dependencies for each appraisal:
`bundle exec appraisal install`

Run `active_record_6` appraisal:
bundle exec appraisal active_record_6 rake spec --backtrace

### CODE REVIEW

File: arel_toolkit.gemspec [hunk 1]
```diff
@@ -24,9 +24,8 @@ Gem::Specification.new do |spec|
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ['lib']
   spec.extensions    = ['ext/pg_result_init/extconf.rb']
-
-  spec.add_dependency 'arel', '~> 9.0.0'
```
Comment (khaleksa) [inline arel_toolkit.gemspec:28]: 'arel' was moved to 'activerecord' in Rails 6

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -22,6 +22,7 @@
 
 RSpec.configure do |config|
   # Enable flags like --only-failures and --next-failure
+  config.full_backtrace = true
```
Comment (khaleksa) [inline spec/spec_helper.rb]: Needed for debugging, should be removed before merging

File: Appraisals [hunk 3]
```diff
@@ -7,3 +7,8 @@ end
 
 appraise 'default' do
 end
+
+appraise 'active_record_6' do
+  gem 'activerecord', '~> 6.0.0'
+  gem 'pg_query', '~> 1.2.0'
```
Comment (mvgijssel) [inline Appraisals]: Why is the pg_query gem upgrade here as well? Shouldn't this be part of the gemspec?

File: arel_toolkit.gemspec (hunk 1 above)
Comment (mvgijssel) [reply arel_toolkit.gemspec:28]: So interestingly enough you've removed arel from the gemspec, didn't add it to the Appraisal and everything still works? 🤔 . Does that mean Arel is required by ActiveRecord in version 5 and 6?

File: lib/arel/sql_to_arel/pg_query_visitor.rb [hunk 4]
```diff
@@ -719,7 +719,9 @@ def visit_SelectStmt(
         select_core.havings = [visit(having_clause)] if having_clause
         select_core.windows = visit(window_clause) if window_clause
         select_core.into = visit(into_clause) if into_clause
-        select_core.top = ::Arel::Nodes::Top.new visit(limit_count) if limit_count
+        if (ActiveRecord.version.version.to_i != 6) && limit_count
```
Comment (mvgijssel) [inline lib/arel/sql_to_arel/pg_query_visitor.rb]: Why is this change necessary? If this is because `::Arel::Nodes::Top` does not exist, you might as well define it in `lib/arel/extensions/top.rb`

**Review by mvgijssel — COMMENTED** [review]
Can you add a spec to `spec/arel_toolkit_spec.rb` which checks the Arel::VERSION? For AR 5 it should be 9 and for AR 6 it should be 10.

This checks if actually the right Arel library is loaded.

File: arel_toolkit.gemspec (hunk 1 above)
Comment (khaleksa) [reply arel_toolkit.gemspec:28]: Your assumption is right, `arel` is in runtime dependencies of ActiveRecord
https://rubygems.org/gems/activerecord/versions/5.0.0.1

File: Appraisals (hunk 3 above)
Comment (khaleksa) [reply Appraisals]: Right, I will move it to gemspec!

File: lib/arel/sql_to_arel/pg_query_visitor.rb (hunk 4 above)
Comment (khaleksa) [reply lib/arel/sql_to_arel/pg_query_visitor.rb]: This block is necessary for active_record 5.2, without it Arel tree has a different structure (Arel tree includes `Arel::Nodes::Top` while it should be nil). Please, see the two arel trees in the screenshot

image

or run the spec `rspec ./spec/arel/sql_to_arel_spec.rb:874`

**Comment by khaleksa** [comment]
> Can you add a spec to `spec/arel_toolkit_spec.rb` which checks the Arel::VERSION? For AR 5 it should be 9 and for AR 6 it should be 10.
> 
> This checks if actually the right Arel library is loaded.

done

File: lib/arel/sql_to_arel/pg_query_visitor.rb (hunk 4 above)
Comment (mvgijssel) [reply lib/arel/sql_to_arel/pg_query_visitor.rb]: Yeah I get you, but I'm saying we can keep the Arel::Nodes::Top node so we don't need this conditional branch :).

**Review by mvgijssel — APPROVED** [review]

File: spec/arel/middleware_spec.rb [hunk 5]
```diff
@@ -423,9 +425,10 @@ def self.call(_arel, next_middleware)
     after_sql = []
 
     logger = lambda do |next_arel, next_middleware, context|
+      result = next_middleware.call(next_arel)
       after_sql << next_arel.to_sql
       before_sql << context[:original_sql]
-      next_middleware.call(next_arel)
+      result
```
Comment (Willianvdv) [inline spec/arel/middleware_spec.rb:431]: Interesting that you got you an out of order array of SQLs without this change.

**Review by Willianvdv — APPROVED** [review]

File: lib/arel/sql_to_arel/pg_query_visitor.rb (hunk 4 above)
Comment (khaleksa) [reply lib/arel/sql_to_arel/pg_query_visitor.rb]: Got it! thank you for the suggestion, @mvgijssel

Comment (khaleksa) [reply lib/arel/sql_to_arel/pg_query_visitor.rb]: I tried to define `Top` in `lib/arel/extensions/top.rb`
```
# frozen_string_literal: true

module Arel # :nodoc: all
  module Nodes
    class Top < Unary
    end
  end
end
```
As result, all specs passed for ActiveRecord 5.2 but some specs failed for ActiveRecord 6.0 with an error `message: uninitialized constant Arel::Nodes::Top`.

@mvgijssel  Given the time pressure, I would like to go like it is now if you don't have objection?

Comment (khaleksa) [reply lib/arel/sql_to_arel/pg_query_visitor.rb]: The issue was fixed after adding reference to file. @mvgijssel thanks for the suggestion!

File: spec/arel/middleware_spec.rb (hunk 5 above)
Comment (khaleksa) [reply spec/arel/middleware_spec.rb:431]: it is even more interesting why the old code works with arel 9
